Update to rulesnodejs@3.1.0

josephperrott opened the pull request:
See individual commits.


See [rules_nodejs#stricter-ts-typechecks](https://github.com/bazelbuild/rules_nodejs/wiki#stricter-ts-typechecks) for more information on changes to castings involving `JSON.parse`

josephperrott committed:
build: update to rules_nodejs@3.1.0

Update to the latest version of rules_nodejs.

josephperrott committed:
fix(bazel): update integration test to use rules_nodejs@3.1.0

Update the integraiton tests for bazel to the latest rules_nodejs version.

josephperrott committed:
build: update tools directory for latest changes in rules_nodejs

Update the scripts/tooling in the tools directory to handle the changes in the latest
version of rules_nodejs.

josephperrott committed:
fix(zone.js): update build tooling for latest changes in rules_nodejs

Update the zone.js build tooling to handle the changes in the latest version of rules_nodejs.

josephperrott committed:
fix(service-worker): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(router): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(platform-browser): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(localize): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(elements): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(compiler-cli): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(compiler): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(benchpress): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(bazel): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(dev-infra): update type castings for JSON.parse usage

Update usages of JSON.parse to be cast as specific types.

josephperrott committed:
fix(bazel): update build tooling for latest changes in rules_nodejs

Update the build tooling to handle the changes in the latest version of rules_nodejs.

josephperrott committed:
fix(dev-infra): update build tooling for latest changes in rules_nodejs

Update the build tooling to handle the changes in the latest version of rules_nodejs.

AndrewKushnir reviewed (commented):
@josephperrott could you plz share a link (may be add to a PR description) to a change that triggered JSON.parse type cast (as an additional context for CR)?

alan-agius4 commented:
@AndrewKushnir https://github.com/bazelbuild/rules_nodejs/wiki#stricter-ts-typechecks

alan-agius4 commented on packages/bazel/src/ng_package/packager.ts:
```diff
@@ -91,7 +91,7 @@ function main(args: string[]): number {
   const srcs = srcsArg.split(',').filter(s => !!s);
   const dataFiles: string[] = dataArg.split(',').filter(s => !!s);
-  const modulesManifest = JSON.parse(modulesManifestArg);
+  const modulesManifest = JSON.parse(modulesManifestArg) as {[key: string]: any};
```
```suggestion
  const modulesManifest = JSON.parse(modulesManifestArg) as Record<string, any>;
```

alan-agius4 commented:
@josephperrott, most of the casts are `as any` maybe it's just better to disable the plugin.

josephperrott commented:
@alan-agius4 I updated the PR to include more specific typings throughout.

petebacondarwin commented on packages/compiler-cli/ngcc/test/packages/configuration_spec.ts:
```diff
@@ -567,7 +567,9 @@ runInEachFileSystem(() => {
-        afterEach(() => DEFAULT_NGCC_CONFIG.packages = JSON.parse(originalDefaultConfig));
+        afterEach(
+            () => DEFAULT_NGCC_CONFIG.packages =
+                JSON.parse(originalDefaultConfig) as {[key: string]: RawNgccPackageConfig});
```
How about:

```suggestion
                JSON.parse(originalDefaultConfig) as NgccProjectConfig['packages'];
```

petebacondarwin commented on packages/compiler-cli/ngcc/test/packages/entry_point_spec.ts:
```diff
@@ -116,8 +116,8 @@ runInEachFileSystem(() => {
       const config = new NgccConfiguration(fs, _('/project'));
-      const info: EntryPoint =
-          getEntryPointInfo(fs, config, new MockLogger(), SOME_PACKAGE, entryPointPath) as any;
+      const info = getEntryPointInfo(fs, config, new MockLogger(), SOME_PACKAGE, entryPointPath) as
+          EntryPoint;
```
👍

petebacondarwin commented on packages/compiler-cli/ngcc/test/rendering/renderer_spec.ts:
```diff
@@ -641,7 +641,8 @@ UndecoratedBase.ɵdir = ɵngcc0.ɵɵdefineDirective({ type: UndecoratedBase, vie
                  .toEqual(RENDERED_CONTENTS + '\n' + generateMapFileComment('file.js.map'));
              expect(mapFile.path).toEqual(_('/node_modules/test-package/src/file.js.map'));
-             expect(JSON.parse(mapFile.contents)).toEqual(MERGED_OUTPUT_PROGRAM_MAP.toObject());
+             expect(JSON.parse(mapFile.contents) as SourceMapConverter)
+                 .toEqual(MERGED_OUTPUT_PROGRAM_MAP.toObject());
```
This one really should be `as any`, since that is the return type of `SourceMapConverter.toObject()`.

```suggestion
             expect(JSON.parse(mapFile.contents) as any)
                 .toEqual(MERGED_OUTPUT_PROGRAM_MAP.toObject());
```

petebacondarwin commented on packages/compiler-cli/ngcc/test/writing/new_entry_point_file_writer_spec.ts:
```diff
@@ -184,7 +184,9 @@ runInEachFileSystem(() => {
 
-           expect(JSON.parse(fs.readFile(_('/node_modules/test/__ivy_ngcc__/es2015/index.js.map'))))
+           expect(
+               JSON.parse(fs.readFile(_('/node_modules/test/__ivy_ngcc__/es2015/index.js.map'))) as
```
I think we could use import `RawSourceMap` from `'../../../src/ngtsc/sourcemaps'` here, and below.

josephperrott replied:
It looks like the typings of those don't line up.

```ts
export interface RawSourceMap {
  version: number|string;
  file?: string;
  sourceRoot?: string;
  sources: string[];
  names: string[];
  sourcesContent?: (string|null)[];
  mappings: string;
}
```
where as the object we are checking comes out to

```ts
interface {
  version: number;
  file: string;
  sources: string[];
  mappings: string;
  sourceRoot: string;
}
```

petebacondarwin replied:
That actually highlights an issue with the test data :-)
Earlier one we have 

```ts
const sourceMap = {
  version: 3,
  file: 'index.js',
  sources: ['../src/index.ts'],
  mappings: 'AAAA',
};
```

This is actually wrong and should be:

```ts
const sourceMap: RawSourceMap = {
  version: 3,
  file: 'index.js',
  sources: ['../src/index.ts'],
  names: [],
  mappings: 'AAAA',
};
```

petebacondarwin commented on packages/compiler-cli/ngcc/test/writing/new_entry_point_file_writer_spec.ts:
```diff
@@ -678,7 +682,8 @@ runInEachFileSystem(() => {
             ],
             ['fesm5', 'module']);
-        const packageJsonFromFile1 = JSON.parse(fs.readFile(packageJsonPath));
+        const packageJsonFromFile1 = JSON.parse(fs.readFile(packageJsonPath)) as
+            {fesm5_ivy_ngcc?: string, module_ivy_ngcc?: string};
```
OMG in `ngcc/src/packages/entry_points.ts` we should create the following type:

```ts
export type NewEntryPointPropertiesMap = {
  [Property in PackageJsonFormatProperties as `${Property}_ivy_ngcc`]?: string;
};
```

And then this can be:

```suggestion
        const packageJsonFromFile1 = JSON.parse(fs.readFile(packageJsonPath)) as
            NewEntryPointPropertiesMap;
```

petebacondarwin commented on packages/compiler-cli/ngcc/test/writing/package_json_updater_spec.ts:
```diff
@@ -18,7 +18,7 @@ runInEachFileSystem(() => {
 
     // Helpers
-    const readJson = (path: AbsoluteFsPath) => JSON.parse(fs.readFile(path));
+    const readJson = <T>(path: AbsoluteFsPath) => JSON.parse(fs.readFile(path)) as T;
```
🎉

petebacondarwin replied:
Although.... in this case we never actually access any of the properties of the parsed objects directly. So I think it might be simpler (for the tests) just to go with `any` here!

petebacondarwin commented on packages/compiler-cli/src/transformers/metadata_reader.ts:
```diff
@@ -70,7 +70,8 @@ function readMetadataFile(host: MetadataReaderHost, dtsFilePath: string): Module
   try {
-    const metadataOrMetadatas = JSON.parse(host.readFile(metadataPath));
+    const metadataOrMetadatas =
+        JSON.parse(host.readFile(metadataPath)) as ModuleMetadata[] | undefined;
```
I think (to be exact):

```suggestion
        JSON.parse(host.readFile(metadataPath)) as ModuleMetadata | ModuleMetadata[] | undefined;
```

petebacondarwin replied:
And then we can drop the type on the `metadatas` const below.

petebacondarwin commented on packages/compiler-cli/test/compliance/partial/partial_compliance_goldens.bzl:
```diff
@@ -21,6 +21,7 @@ def partial_compliance_golden(filePath):
         entry_point = "//packages/compiler-cli/test/compliance/partial:cli.ts",
         templated_args = [
             filePath,
+            "--bazel_patch_module_resolver",
```
Ooh! What effect does this have?
And will it allow us to generate golden files on Windows??

josephperrott replied:
I updated the PR to note it needs to be removed, but yes it puts the monkey patching back in for the `resolve` tooling to work with bazel.  We need to go back and fixup all of the dependencies in the build rules to be able to remove it again.

petebacondarwin commented on packages/compiler-cli/test/compliance/test_helpers/get_compliance_tests.ts:
```diff
@@ -55,9 +55,10 @@ export function* getComplianceTests(testConfigPath: string): Generator<Complianc
+    fs: ReadonlyFileSystem, testCasesPath: AbsoluteFsPath, basePath: AbsoluteFsPath) {
   try {
-    return JSON.parse(fs.readFile(testCasesPath));
+    return JSON.parse(fs.readFile(testCasesPath)) as
+        {cases: {expectations: Expectation[], focusTest: boolean, excludeTest: boolean}[]};
```
This is not quite accurate. The parsed value can be a single object (say `TestCaseJson`) or an array of objects (`TestCaseJson[]`), where `TestCaseJson` is:

```ts
interface TestCaseJson {
  description: string;
  compilationModeFilter?: ('fulll compile'|'linked compile')[];
  inputFiles?: string[];
  expectations?: {
    failureMessage?: string;
    files?: ExpectedFiles[]|string;
    expectedErrors?: {message: string, location?: string};
    extraChecks?: (string|string[])[];
  };
  compilerOptions?: ConfigOptions>;
  angularCompilerOptions?: ConfigOptions>;
  focusTest?: boolean;
  excludeTest?: boolean;
}
```

I think we should probably create this interface and use it.
(By the way it has to match the definition in `packages/compiler-cli/test/compliance/test_cases/test_case_schema.json` so that might be worth a comment.

petebacondarwin commented on packages/elements/schematics/ng-add/index_spec.ts:
```diff
@@ -58,7 +58,7 @@ describe('Elements Schematics', () => {
         await schematicRunner.runSchematicAsync('ng-add', defaultOptions, appTree).toPromise();
     const pkgJsonText = tree.readContent('/package.json');
-    const pkgJson = JSON.parse(pkgJsonText);
+    const pkgJson = JSON.parse(pkgJsonText) as {dependencies: any};
```
I feel like `@angular-devkit/schematics/testing` should export some `PackageJsonObject` interface to be used by all these schematics tests, since loading a package.json is so common.

petebacondarwin commented on packages/localize/src/tools/src/extract/translation_files/format_options.ts:
```diff
@@ -40,5 +40,5 @@ export function validateOptions(name: string, validOptions: ValidOptions, option
-}
\ No newline at end of file
+  return JSON.parse(optionString) as FormatOptions;
+}
```
❤️

petebacondarwin commented on packages/localize/src/tools/src/translate/translation_files/translation_parsers/simple_json_translation_parser.ts:
```diff
@@ -46,7 +46,7 @@ export class SimpleJsonTranslationParser implements TranslationParser<Object> {
     }
     try {
-      const json = JSON.parse(contents);
+      const json = JSON.parse(contents) as ParsedTranslationBundle;
```
This is not quite right since `ParsedTranslationBundle` contains a required `diagnostics: Diagnostics` property, which is not in the JSON.

Instead we should create a local interface in this file:

```ts
interface SimpleJsonFile {
  locale: string;
  translations: {[messageId: string]: string};
}
```

and use it here and also on line 74 below.

petebacondarwin commented on packages/localize/src/tools/src/translate/translation_files/translation_parsers/simple_json_translation_parser.ts:
```diff
@@ -71,7 +71,7 @@ export class SimpleJsonTranslationParser implements TranslationParser<Object> {
 
   parse(_filePath: string, contents: string, json?: Object): ParsedTranslationBundle {
-    const {locale: parsedLocale, translations} = json || JSON.parse(contents);
+    const {locale: parsedLocale, translations} = json || JSON.parse(contents) as any;
```
```suggestion
    const {locale: parsedLocale, translations} = (json || JSON.parse(contents)) as SimpleJsonFile;
```

petebacondarwin commented on tools/symbol-extractor/symbol_extractor.ts:
```diff
@@ -74,14 +74,14 @@ export class SymbolExtractor {
   static diff(actual: Symbol[], expected: string|((Symbol | string)[])): {[name: string]: number} {
     if (typeof expected == 'string') {
-      expected = JSON.parse(expected);
+      expected = JSON.parse(expected) as (Symbol | string)[];
```
I don't think that JSON.parse can produce `Symbol` objects so...

```suggestion
      expected = JSON.parse(expected) as string[];
```

petebacondarwin commented on tools/symbol-extractor/symbol_extractor.ts:
```diff
@@ -74,14 +74,14 @@ export class SymbolExtractor {
     // with that name. Once they are matched with symbols in the actual output, the count should
     // even out to 0.
-    (expected as (Symbol | string)[]).forEach((nameOrSymbol) => {
+    expected.forEach((nameOrSymbol) => {
```
👍 
You can even remove the parens...

```suggestion
    expected.forEach(nameOrSymbol => {
```

petebacondarwin reviewed (changes requested):
Awesome (and wide ranging!!) PR @josephperrott.

I made a whole host of suggestions for the casts; primarily focussed on areas that I have some ownership over (e.g. ngcc and localize).
I skipped over some of the other packages like elements and router, etc.

josephperrott commented on WORKSPACE:
```diff
@@ -31,6 +31,7 @@ yarn_install(
     name = "npm",
     manual_build_file_contents = npm_package_archives(),
     package_json = "//:package.json",
+    strict_visibility = False,
```
It is related to strict dependencies of npm packages and their presence in the package.json file, I added a comment in the WORKSPACE file to express that a follow up is needed to correct this.  It just will be a ton of tiny changes everywhere.

josephperrott commented on dev-infra/release/publish/actions.ts:
```diff
@@ -82,7 +82,8 @@ export abstract class ReleaseAction {
     const pkgJsonPath = join(this.projectDir, packageJsonPath);
-    const pkgJson = JSON.parse(await fs.readFile(pkgJsonPath, 'utf8'));
+    const pkgJson =
+        JSON.parse(await fs.readFile(pkgJsonPath, 'utf8')) as {version: string, [key: string]: any};
```
I think a helper function for parsing JSON would be helpful, but I don't know that it makes sense to do at this point.

I believe that `readFileSync` comes from `fs-extra` rather than `fs` so we avoided using it.

petebacondarwin replied:
`readFileSync` is straight from node.js: https://nodejs.org/dist/latest-v14.x/docs/api/fs.html#fs_fs_readfilesync_path_options

petebacondarwin replied:
But I guess that you are consistently using async functions throughout so it makes sense to leave this as-is.

petebacondarwin commented on packages/compiler-cli/ngcc/test/writing/new_entry_point_file_writer_spec.ts:
```diff
@@ -5,13 +5,15 @@
  * Use of this source code is governed by an MIT-style license that can be
  * found in the LICENSE file at https://angular.io/license
  */
+import {RawSourceMap} from '@angular/compiler-cli/src/ngtsc/sourcemaps';
```
```suggestion
import {RawSourceMap} from '../../../src/ngtsc/sourcemaps';
```

petebacondarwin reviewed (approved):
NICE!

IgorMinar commented on .circleci/config.yml:
```diff
@@ -782,98 +782,10 @@ workflows:
           filters:
             branches:
               ignore: g3
-      - lint:
```
What are all these CI changes for The commit message just says "WIP". Is this to be removed?

josephperrott replied:
Yes, I am working to solve the windows test failures, and need to run the test in a windows environment.  The will be removed.

IgorMinar replied:
oh, I see. thanks

IgorMinar reviewed (changes requested):
lgtm except for the CI change. can you please clarify what's going on there? thanks!

IgorMinar reviewed (approved):
LGTM! Thank you!

Reviewed-for: global-approvers

josephperrott committed:
fix(compiler-cli): update ngcc integration tests for latest changes in rules_nodejs

Update the ngcc integration tests to handle the changes in the latest version of rules_nodejs.

josephperrott commented:
Caretaker Note: We should merge this and then watch the next few merges after to make sure there are not flakes.
